crates/sncast/src/starknet_commands/script.rs

```diff
@@ -38,14 +40,182 @@ use sncast::helpers::configuration::CastConfig;
 use sncast::helpers::constants::SCRIPT_LIB_ARTIFACT_NAME;
 use sncast::response::print::print_as_warning;
 use sncast::response::structs::ScriptResponse;
+use sncast::{ErrorData, TransactionError};
 use starknet::accounts::Account;
+use starknet::core::types::StarknetError;
 use starknet::core::types::{BlockId, BlockTag::Pending, FieldElement};
 use starknet::providers::jsonrpc::HttpTransport;
-use starknet::providers::JsonRpcClient;
+use starknet::providers::{JsonRpcClient, ProviderError};
 use tokio::runtime::Runtime;
 
 type ScriptStarknetContractArtifacts = StarknetContractArtifacts;
 
+pub trait SerializeAsFelt252 {
+    fn serialize_as_felt252(&self) -> Vec<Felt252>;
+}
+
+#[derive(Debug)]
+enum ScriptCommandError {
+    ContractArtifactsNotFound(ErrorData),
+    ScriptProviderError(ScriptProviderError),
+}
+
+impl From<RecoverableStarknetCommandError> for ScriptCommandError {
+    fn from(value: RecoverableStarknetCommandError) -> Self {
+        match value {
+            RecoverableStarknetCommandError::ContractArtifactsNotFound(err_data) => {
+                ScriptCommandError::ContractArtifactsNotFound(err_data)
+            }
+            RecoverableStarknetCommandError::TransactionError(error) => match error {
+                TransactionError::Rejected => ScriptCommandError::ScriptProviderError(
+                    ScriptProviderError::StarknetError(ScriptStarknetError::TransactionRejected),
+                ),
+                TransactionError::Reverted(err_data) => {
+                    ScriptCommandError::ScriptProviderError(ScriptProviderError::StarknetError(
+                        ScriptStarknetError::TransactionReverted(err_data),
+                    ))
+                }
+            },
+            RecoverableStarknetCommandError::ProviderError(error) => {
+                ScriptCommandError::ScriptProviderError(error.into())
+            }
+        }
+    }
+}
```

Comment:
I wonder if it would make sense to try and refactor impls for subcommands + script errors to combine such checks together. This particular one looks boilerplate-ish to me, since it basically covers the same errors as subcommands.

Reply:
My idea was for `ScriptCommandError` and `StarknetCommandError` to be independent from one another, so that `ScriptCommandError` would have a structure that makes the most sense from the user's perspective, and  `StarknetCommandError` a structure that makes the most sense from "sncast internals" perspective

Reply:
Also to have a clear separation between `ScriptCommandError` and `StarknetCommandError`, so they can evolve independently

Reply:
> This particular one looks boilerplate-ish to me, since it basically covers the same errors as subcommand.

I see what you are talking about, but I'm not sure how we can improve this

Reply:
Ok, so let's just sum up current behaviour:

- we have recoverable/unrecoverable errors for script
- we do not have such distinction for plain subcommands, we just throw an err basically
- we return different types for buth scenarios, to distinct `subcommands errors` and `script errors`, which under the hood are basically the same right?

do you think it would make sense to just have `SncastError` struct that will be used for both script and subcommands, and that would not distinct recoverable/unrecoverable? at the end of the day, we use the same functions for subcommands and script to declare/deploy etc, and I guess it would make sense to allow users handle network errors etc (unrecoverable ones basically) in script - that would allow to write sturdier scripts in general in my opinion. So it would be a win for users, and a win for us, as it would bring down the code complexity.

thoughts? am I missing something here?

Reply:
After an internal discussion, we decided to use StarknetCommandError for scripts, but restrict `StarknetCommandError::ProviderError::StarknetError` to only those variants that may occur (some will not since we use only selected rpc methods)